### PR TITLE
Feat(eos_cli_config_gen): Add Trident MMU queue

### PR DESCRIPTION
**Pull request by chetryan**
Add the support for defining queue settings for the Memory Management Unit on Trident platforms.

## Change Summary

At the moment, changing the default profile for the **Memory Management Unit (MMU)** on Trident platforms is only possible via `eos_cli`. 
This change adds the functionality to `eos_cli_config_gen` to accept this as part of the data model.

## Related Issue(s)

No current Issues raised.

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes

- Add functionality to the `eos/platform.j2` to generate device config
- Add functionality to the `documentation/platform.j2` to generate matching Markdown docs for devices
- Extend the schema to accept the dictionary for mmu settings

## How to test

- Run `molecule test -s eos_cli_config_gen`

### User Checklist

- [] Review generated `platform.md` and give feedback whether this is clear enough.
- [] Review the generated configuration `platform.cfg` and comment on any improvements required

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
     "name": "AVD Development",
     "build": {
         "dockerfile": "Dockerfile",
-        "args": { "VARIANT": "3.8-v2.0" },
```
Comment (ClausHolbechArista) [inline .devcontainer/devcontainer.json]: Unrelated change

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg [hunk 2]
```diff
@@ -5,6 +5,28 @@ transceiver qsfp default-mode 4x10G
...
+    egress multicast queue 0 drop-precedence 2 drop-threshold 1
+    egress multicast queue 1 reserved cells 0
+    egress multicast queue 1 threshold 8
+    egress unicast queue 1 reserved bytes 0
+    egress unicast queue 1 drop-precedence 1 drop-threshold 1/8
+    egress unicast queue 2 reserved cells 0
+    egress unicast queue 2 threshold 1/8
 platform sand qos map traffic-class 0 to network-qos 0
 platform sand qos map traffic-class 1 to network-qos 7
 platform sand qos map traffic-class 2 to network-qos 15
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: Are we sure the ordering and spacing is similar to EOS here? It looks wrong to me.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/platform.j2 [hunk 3]
```diff
@@ -12,8 +12,30 @@
 | Settings | Value |
 | -------- | ----- |
 {%             for tridentsetting in platform.trident | arista.avd.natural_sort %}
+{%                 if platform.trident[tridentsetting] is not mapping %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/platform.j2]: please rework this to detect the relevant keys directly instead of this.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2 [hunk 4]
```diff
@@ -4,6 +4,43 @@
...
+{%     for profile in platform.trident.mmu.queue_profiles | arista.avd.natural_sort('name') %}
+!
+platform trident mmu queue profile {{ profile.name }}
+{%         for queue in profile.multicast_queues | arista.avd.natural_sort('id') %}
+{%             if queue.reserved is arista.avd.defined %}
+{%                 set unit = "" %}
+{%                 if queue.unit is arista.avd.defined %}
+{%                     set unit = " " ~ queue.unit | arista.avd.default('') %}
+{%                 endif %}
+    egress multicast queue {{ queue.id }} reserved{{ unit }} {{ queue.reserved }}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2]: I would prefer to use the same style as we do elsewhere:
```suggestion
{%             if queue.reserved is arista.avd.defined %}
{%                 set reserved_cli = "egress multicast queue " ~ queue.id ~ " reserved" %}
{%                 if queue.unit is arista.avd.defined %}
{%                     set reserved_cli = reserved_cli ~ " " ~ queue.unit %}
{%                 endif %}
{%                 set reserved_cli = reserved_cli ~ " " ~ queue.reserved %}
    {{ reserved_cli }}
```

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (chetryan) [reply .devcontainer/devcontainer.json]: Agreed - this was related to my local environment. Reverting

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg (hunk 2 above)
Comment (chetryan) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: Tested on a 7050DX and it looks to me similar to what is being presented in the MR. Do you see any specific differences with regards to spacing? I will fix the ordering to have unicast first.

image

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2 (hunk 4 above)
Comment (chetryan) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2]: Agreed, updated the template.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg (hunk 2 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: You need to run without any section commands. You have `!` between each profile. Maybe that is fine - I don't know. Also at the end in the example I commented on, where is no divider to the next section (which may be a different problem).
Also your output here shows that the ordering of multicast and unicast is opposite of the PR.
Please test with the full platform command section from the molecule config.

Comment (chetryan) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: Below is the snippet from show run (no section command).
- the output has ! between each profile
- in reality, would we ever have the trident and sand platform being referenced on the same device? If I add a ! separator to between sand and trident, I need to add logic to make sure we don't get two ! consecutively if there is not `platform trident` section. Shall I do that?
- Agreed with the multicast/unicast ordering and that should be fixed now
- I have already updated the platform command section of the molecule config. Not sure what else to add/test
 
image

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2 [hunk 5]
```diff
@@ -4,6 +4,48 @@
...
+{%             if queue.threshold is arista.avd.defined %}
+    egress multicast queue {{ queue.id }} threshold {{ queue.threshold }}
+{%             endif %}
+{%             if queue.drop is arista.avd.defined %}
+    egress multicast queue {{ queue.id }} drop-precedence {{ queue.drop.precedence }} drop-threshold {{ queue.drop.threshold }}
+{%             endif %}
+{%         endfor %}
+{%         if loop.last %}
+!
+{%         endif %}
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2]: We normally never insert these at the _end_ of a section. I think it might be missing from the following section, which causes the issue I commented on.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg [hunk 6]
```diff
@@ -5,6 +5,29 @@ transceiver qsfp default-mode 4x10G
 hostname platform
 !
 platform trident forwarding-table partition 2
+platform trident mmu queue profile mc_example_profile apply
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: when I apply this on a dut running 4.30, this command comes at the very end of the config.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg [hunk 7]
```diff
@@ -5,6 +5,29 @@ transceiver qsfp default-mode 4x10G
...
+!
+platform trident mmu queue profile mc_example_profile
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg:5]: When I apply this on a dut running 4.30, the profiles come right before the hostname.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/platform.j2 [hunk 8]
```diff
@@ -11,9 +11,29 @@
...
-{%             endfor %}
+{%             if platform.trident.forwarding_table_partition is arista.avd.defined %}
+| Forwarding Table Partition | {{ platform.trident.forwarding_table_partition }} |
+{%             endif %}
+{%             if platform.trident.mmu.active_profile is arista.avd.defined %}
+| MMU Applied Profile | {{ platform.trident.mmu.active_profile }} |
+{%             endif %}
+{%             if platform.trident.mmu.queue_profiles is arista.avd.defined %}
+
+#### Trident MMU QUEUE PROFILES**
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/platform.j2]:
```suggestion
#### Trident MMU QUEUE PROFILES
```

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg (hunk 7 above)
Comment (chetryan) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg:5]: Agreed, changed

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg (hunk 6 above)
Comment (chetryan) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: Agreed, changed the template accordingly.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2 (hunk 5 above)
Comment (chetryan) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/platform.j2]: understood, removed from the template with the refactor.

File: ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg (hunk 2 above)
Comment (chetryan) [reply ansible_collections/arista/avd/molecule/eos_cli_config_gen/intended/configs/platform.cfg]: This should now be resolved with the refactor of the Jinja template.

**Review by ClausHolbechArista — APPROVED** [review]

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/platform.schema.yml [hunk 9]
```diff
@@ -13,6 +13,119 @@ keys:
...
+                        type: dict
+                        keys:
+                          id:
+                            type: int
+                            convert_types:
+                              - str
+                            required: true
+                            min: 0
+                            max: 7
+                          unit:
```
Comment (carlbuchmann) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/platform.schema.yml:50]: to make this more intuitive I would suggest to rename to `reserved_unit`

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/platform.schema.yml [hunk 10]
```diff
@@ -13,6 +13,119 @@ keys:
...
```
Comment (carlbuchmann) [inline ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/platform.schema.yml:97]: to make this more intuitive I would suggest to rename to reserved_unit

**Review by carlbuchmann — APPROVED** [review]
Reviewed and tested on DUT, with minor comments on the schema. I will also approve as is.

File: ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/platform.schema.yml (hunk 9 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_cli_config_gen/schemas/schema_fragments/platform.schema.yml:50]: I would expect this unit to also impact the unit of the thresholds, but I don't know. We should verify on EOS.
